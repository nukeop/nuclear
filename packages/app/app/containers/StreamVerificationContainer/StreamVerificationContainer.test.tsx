import { render } from '@testing-library/react';
import React from 'react';
import fetchMock from 'fetch-mock';

import { AnyProps } from '../../../test/testUtils';
import { StreamVerificationContainer } from '.';
import { buildStoreState } from '../../../test/storeBuilders';
import { configureMockStore, TestStoreProvider, setupI18Next } from '../../../test/testUtils';

describe('StreamVerificationContainer', () => {
  beforeAll(() => {
    process.env.NUCLEAR_VERIFICATION_SERVICE_URL = 'http://verification.nuclear';
    setupI18Next();
  });

  beforeEach(() => {
    fetchMock.reset();
  });
      
  it('renders nothing when nothing\'s playing', () => {
    const { component } = mountComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('renders the current status when there\'s a track in the queue', () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', {
      stream_id: 'stream-id',
      score: 10
    });
    const store = buildStoreState()
      .withTracksInPlayQueue()
      .withSettings({
        isReady: true,
        useStreamVerification: true
      })
      .build();
    const { component } = mountComponent(store);
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('shows the track as weakly verified if there are few verifications', async () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', {
      stream_id: 'CuklIb9d3fI',
      score: 1
    });
    const store = buildStoreState()
      .withTracksInPlayQueue()
      .withSettings({
        isReady: true,
        useStreamVerification: true
      })
      .build();
    const { component } = mountComponent(store);
    await component.findByText('Weakly verified');
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('shows the track as verified if there are many verifications', async () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', {
      stream_id: 'CuklIb9d3fI',
      score: 10
    });
    const store = buildStoreState()
      .withTracksInPlayQueue()    
      .withSettings({
        isReady: true,
        useStreamVerification: true
      })
      .build();
    const { component } = mountComponent(store);
    await component.findByText('Verified');
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('shows the track as verified by user if the user has verified it', async () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', {
      stream_id: 'CuklIb9d3fI',
      score: 10,
      self_verified: true
    });
    const store = buildStoreState()
      .withTracksInPlayQueue()
      .withSettings({
        userId: '1',
        useStreamVerification: true,
        isReady: true
      })
      .build();

    const { component } = mountComponent(store);
    await component.findByText('Verified by you');
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('lets the user verify the track', async () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', 404);
    fetchMock.post({ url: 'http://verification.nuclear/stream-mappings/verify', body: {
      stream_id: 'CuklIb9d3fI',
      artist: 'test artist 1',
      title: 'test track 1',
      source: 'Test Stream Provider',
      author_id: '1'
    }}, 201);

    const store = buildStoreState()
      .withTracksInPlayQueue()
      .withPlugins()
      .withSettings({
        userId: '1',
        useStreamVerification: true,
        isReady: true
      })
      .build();
    const { component } = mountComponent(store);
    await (await component.findByText('Verify')).click();

    expect(fetchMock.done()).toBeTruthy();
  });

  it('lets the user unverify the track', async () => {
    fetchMock.post('http://verification.nuclear/stream-mappings/top-stream', {
      stream_id: 'CuklIb9d3fI',
      score: 10,
      self_verified: true
    });
    fetchMock.delete({
      url: 'http://verification.nuclear/stream-mappings/unverify',
      body: {
        stream_id: 'CuklIb9d3fI',
        artist: 'test artist 1',
        title: 'test track 1',
        source: 'Test Stream Provider',
        author_id: '1'
      }
    }, 204);
    const store = buildStoreState()
      .withTracksInPlayQueue()
      .withPlugins()
      .withSettings({
        userId: '1',
        useStreamVerification: true,
        isReady: true
      })
      .build();
    const { component } = mountComponent(store);
    
    await (await component.findByText('Unverify')).click();

    expect(fetchMock.done()).toBeTruthy();
  });

  const mountComponent = (initialStore?: AnyProps) => {
    const initialState = initialStore ||
      buildStoreState()
        .withSettings({
          isReady: true,
          useStreamVerification: true
        })
        .build();
    const store = configureMockStore(initialState);
    const component = render(
      <TestStoreProvider
        store={store}
      >
        <StreamVerificationContainer />
      </TestStoreProvider>
    );

    return { component, store };
  };
});
