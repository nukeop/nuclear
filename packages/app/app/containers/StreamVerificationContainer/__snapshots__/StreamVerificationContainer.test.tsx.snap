// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamVerificationContainer renders nothing when nothing's playing 1`] = `<DocumentFragment />`;

exports[`StreamVerificationContainer renders the current status when there's a track in the queue 1`] = `
<DocumentFragment>
  <div
    class="stream_verification"
  >
    <div
      class="stream_verification_right"
    >
      <div
        class="stream_status_icon unknown"
      >
        <i
          aria-hidden="true"
          class="check circle icon"
        />
      </div>
      <div
        class="stream_status"
      >
        Unknown
      </div>
    </div>
    <button
      class="ui button nuclear button"
    >
      Verify
    </button>
  </div>
</DocumentFragment>
`;

exports[`StreamVerificationContainer shows the track as verified by user if the user has verified it 1`] = `
<DocumentFragment>
  <div
    class="stream_verification"
  >
    <div
      class="stream_verification_right"
    >
      <div
        class="stream_status_icon verified_by_user"
      >
        <i
          aria-hidden="true"
          class="check circle icon"
        />
      </div>
      <div
        class="stream_status"
      >
        Verified by you
      </div>
    </div>
    <button
      class="ui button nuclear button"
    >
      Unverify
    </button>
  </div>
</DocumentFragment>
`;

exports[`StreamVerificationContainer shows the track as verified if there are many verifications 1`] = `
<DocumentFragment>
  <div
    class="stream_verification"
  >
    <div
      class="stream_verification_right"
    >
      <div
        class="stream_status_icon verified"
      >
        <i
          aria-hidden="true"
          class="check circle icon"
        />
      </div>
      <div
        class="stream_status"
      >
        Verified
      </div>
    </div>
    <button
      class="ui button nuclear button"
    >
      Verify
    </button>
  </div>
</DocumentFragment>
`;

exports[`StreamVerificationContainer shows the track as weakly verified if there are few verifications 1`] = `
<DocumentFragment>
  <div
    class="stream_verification"
  >
    <div
      class="stream_verification_right"
    >
      <div
        class="stream_status_icon weakly_verified"
      >
        <i
          aria-hidden="true"
          class="check circle icon"
        />
      </div>
      <div
        class="stream_status"
      >
        Weakly verified
      </div>
    </div>
    <button
      class="ui button nuclear button"
    >
      Verify
    </button>
  </div>
</DocumentFragment>
`;
